<a href="https://colab.research.google.com/github/swopnimghimire-123123/Maths_For_ML/blob/main/10_Different_Testing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Notebook 10: Two-Sample Tests & Chi-Square Test
Learning Goals

Compare means between two groups (independent or paired).

Understand t-tests for two samples.

Learn Chi-Square test for categorical data.

Apply in real-world datasets.

# 1. Two-Sample t-Test

`Essence: Compare the means of two groups.`

- Independent t-test: Groups are separate (e.g., boys vs girls exam scores).

- Paired t-test: Same group measured twice (e.g., before & after training).

- Null Hypothesis (H₀): Means are equal.
- Alternative Hypothesis (H₁): Means are different.

In [ ]:
from scipy import stats

# Independent sample example: exam scores
group1 = [85, 88, 90, 92, 87]
group2 = [78, 82, 79, 85, 80]

t_stat, p_value = stats.ttest_ind(group1, group2)
print("t-statistic:", t_stat)
print("p-value:", p_value)

if p_value < 0.05:
    print("Reject H0: Means are different")
else:
    print("Fail to reject H0: Means are similar")


t-statistic: 4.387862045841161
p-value: 0.0023241881225952343
Reject H0: Means are different


### 2. Paired t-Test

- Used when measurements are paired.

- Example: Students’ scores before and after coaching.

In [ ]:
before = [70, 72, 68, 75, 69]
after = [78, 80, 75, 82, 77]

t_stat, p_value = stats.ttest_rel(before, after)
print("t-statistic:", t_stat)
print("p-value:", p_value)

t-statistic: -31.026870075253587
p-value: 6.429804590173413e-06


### 3. Chi-Square Test (Categorical Data)

Essence: Test if two categorical variables are independent.

Null Hypothesis (H₀): Variables are independent.

Alternative Hypothesis (H₁): Variables are related.

In [ ]:
import pandas as pd
from scipy.stats import chi2_contingency

# Example: Gender vs Preference for a product
data = pd.DataFrame({
    'Male': [20, 30],   # Likes, Dislikes
    'Female': [25, 25]  # Likes, Dislikes
}, index=['Likes', 'Dislikes'])

chi2, p, dof, expected = chi2_contingency(data)
print("Chi-square:", chi2)
print("p-value:", p)
print("Expected frequencies:\n", expected)

if p < 0.05:
    print("Reject H0: Gender and preference are related")
else:
    print("Fail to reject H0: Gender and preference seem independent")

Chi-square: 0.6464646464646464
p-value: 0.4213795037428696
Expected frequencies:
 [[22.5 22.5]
 [27.5 27.5]]
Fail to reject H0: Gender and preference seem independent


Here's an explanation of the output from the `chi2_contingency` function:

- **`chi2`**: This is the Chi-Square test statistic. It's a measure of the difference between the observed frequencies in your data and the expected frequencies if the two variables were independent. A larger Chi-Square value suggests a greater difference between observed and expected frequencies.

- **`p`**: This is the p-value. It represents the probability of observing a Chi-Square statistic as extreme as, or more extreme than, the one calculated from your data, assuming the null hypothesis (that the variables are independent) is true. A small p-value (typically less than 0.05) is evidence against the null hypothesis, suggesting the variables are related.

- **`dof`**: This stands for degrees of freedom. It's a value that depends on the size of your contingency table and is used in determining the critical value for the Chi-Square test. For a contingency table with `r` rows and `c` columns, the degrees of freedom are calculated as `(r - 1) * (c - 1)`.

- **`expected`**: This is an array containing the expected frequencies for each cell in the contingency table, assuming the null hypothesis of independence is true. These are the frequencies you would expect to see if there were no relationship between the two variables.

## Problem 1: independent t-test

### Subtask:
Generate sample data for exam scores of two classes and perform an independent t-test to compare their means.


**Reasoning**:
The subtask requires generating sample data for two classes and performing an independent t-test. This can be done in a single code block by generating the lists, performing the t-test, and printing the results with a conclusion based on the p-value.



In [ ]:
import numpy as np
from scipy import stats

# 1. & 2. Generate sample data for exam scores of two classes
np.random.seed(42) # for reproducibility
class_a_scores = np.random.normal(loc=85, scale=5, size=15) # Mean 85, Std Dev 5, 15 scores
class_b_scores = np.random.normal(loc=80, scale=6, size=15) # Mean 80, Std Dev 6, 15 scores

# 3. Perform an independent t-test
t_stat, p_value = stats.ttest_ind(class_a_scores, class_b_scores)

# 4. Print the calculated t-statistic and p-value
print("Independent t-test results:")
print("t-statistic:", t_stat)
print("p-value:", p_value)

# 5. Print a conclusion based on the p-value
alpha = 0.05
if p_value < alpha:
    print(f"With a p-value of {p_value:.4f} and alpha of {alpha}, we reject the null hypothesis.")
    print("Conclusion: The means of the two classes are statistically different.")
else:
    print(f"With a p-value of {p_value:.4f} and alpha of {alpha}, we fail to reject the null hypothesis.")
    print("Conclusion: The means of the two classes are not statistically different.")

Independent t-test results:
t-statistic: 4.187125162923243
p-value: 0.0002539379285855266
With a p-value of 0.0003 and alpha of 0.05, we reject the null hypothesis.
Conclusion: The means of the two classes are statistically different.


## Problem 2: paired t-test

### Subtask:
Generate sample data for blood pressure measurements before and after medication for the same group of people and perform a paired t-test.


**Reasoning**:
Generate sample data for blood pressure measurements before and after medication and perform a paired t-test.



In [ ]:
import numpy as np
from scipy import stats

# 2. Generate blood pressure measurements before medication
np.random.seed(42) # for reproducibility
before_medication = np.random.normal(loc=140, scale=10, size=12) # Mean 140, Std Dev 10, 12 measurements

# 3. Generate blood pressure measurements after medication
# Ensure 'after' measurements are generally lower
after_medication = before_medication - np.random.uniform(low=5, high=15, size=12) # Subtract a random value between 5 and 15

# 4. Perform a paired t-test
t_stat, p_value = stats.ttest_rel(before_medication, after_medication)

# 5. Print the calculated t-statistic and p-value
print("Paired t-test results for blood pressure:")
print("t-statistic:", t_stat)
print("p-value:", p_value)

# 6. Print a conclusion based on the p-value
alpha = 0.05
if p_value < alpha:
    print(f"With a p-value of {p_value:.4f} and alpha of {alpha}, we reject the null hypothesis.")
    print("Conclusion: The medication had a statistically significant effect on blood pressure.")
else:
    print(f"With a p-value of {p_value:.4f} and alpha of {alpha}, we fail to reject the null hypothesis.")
    print("Conclusion: The medication did not have a statistically significant effect on blood pressure.")

Paired t-test results for blood pressure:
t-statistic: 17.262325467097682
p-value: 2.5764052093288428e-09
With a p-value of 0.0000 and alpha of 0.05, we reject the null hypothesis.
Conclusion: The medication had a statistically significant effect on blood pressure.


## Problem 3: chi-square test

### Subtask:
Generate sample data for favorite fruit preference across two age groups and perform a Chi-Square test to see if they are related.


**Reasoning**:
Generate sample data for fruit preference across two age groups and perform a Chi-Square test.



In [ ]:
import pandas as pd
from scipy.stats import chi2_contingency

# 2. Create a pandas DataFrame for observed frequencies
data = pd.DataFrame({
    'Apple': [50, 30],   # Age group 18-30, 31-60
    'Banana': [40, 45],  # Age group 18-30, 31-60
    'Orange': [30, 50]   # Age group 18-30, 31-60
}, index=['18-30', '31-60'])

print("Observed frequencies:")
display(data)

# 3. Perform the Chi-Square test
chi2, p, dof, expected = chi2_contingency(data)

# 4. Print the results
print("\nChi-Square Test Results:")
print("Chi-square statistic:", chi2)
print("p-value:", p)
print("Degrees of freedom:", dof)
print("Expected frequencies:\n", expected)

# 5. Print a conclusion
alpha = 0.05
print("\nConclusion:")
if p < alpha:
    print(f"With a p-value of {p:.4f} and alpha of {alpha}, we reject the null hypothesis.")
    print("There is a significant relationship between age group and fruit preference.")
else:
    print(f"With a p-value of {p:.4f} and alpha of {alpha}, we fail to reject the null hypothesis.")
    print("There is no significant relationship between age group and fruit preference.")

Observed frequencies:


Apple  Banana  Orange
18-30     50      40      30
31-60     30      45      50


Chi-Square Test Results:
Chi-square statistic: 10.196323529411764
p-value: 0.006107964127331901
Degrees of freedom: 2
Expected frequencies:
 [[39.18367347 41.63265306 39.18367347]
 [40.81632653 43.36734694 40.81632653]]

Conclusion:
With a p-value of 0.0061 and alpha of 0.05, we reject the null hypothesis.
There is a significant relationship between age group and fruit preference.


## Summary:

### Data Analysis Key Findings

*   For the independent t-test comparing exam scores of two classes, the p-value (0.0003) was less than the significance level of 0.05, leading to the rejection of the null hypothesis and the conclusion that the means of the two classes are statistically different.
*   The paired t-test on blood pressure measurements before and after medication resulted in a p-value (approximately 2.58e-09) significantly less than 0.05, indicating a statistically significant effect of the medication on blood pressure.
*   The Chi-Square test examining the relationship between age group and fruit preference yielded a p-value (0.0061) less than 0.05, leading to the rejection of the null hypothesis and the conclusion that there is a significant relationship between age group and fruit preference.

### Insights or Next Steps

*   The practice problems demonstrate how to apply different statistical tests (independent t-test, paired t-test, Chi-Square test) to analyze different types of data and draw conclusions based on p-values.
*   The next steps could involve applying these tests to real-world datasets to analyze actual differences in means or relationships between categorical variables.
